### ISSUE TITLE
Input codepoints bigger than 2 bytes (e.g. emojis) get casted into single 2 byte char

### ISSUE THREAD

**Issue by LobbyDivinus** [libgdx/libgdx#7227]
#### Issue details
I noticed in my game that certain characters (i.e. emojis like 🏗️) get not reported correctly to the application via InputProcessor.keyTyped.

What happens in detail:
1. Let's type the emoji 🏗️ which consists of the codepoint 0x1F3D7 (that is, more than two bytes so it will have to be represented using multiple chars)
2. The GLFWCharCallback in https://github.com/libgdx/libgdx/blob/bb799a9cd0a92cadb0425acd6064654cd7a4d6c8/backends/gdx-backend-lwjgl3/src/com/badlogic/gdx/backends/lwjgl3/DefaultLwjgl3Input.java#L51C14-L59C4 receives the correct (4 byte) integer codepoint value.
3. However, the callback casts the codepoint into a (2 byte) and only puts that into the eventQueue for the application to consume. In my example that is 0xF3D7 which is not a valid character.

A solution may be to put the upper 2 bytes of the integer codepoint into the eventQueue, too, if they are non zero.

#### Reproduction steps/code
See description above. All you need is to do is to check the values the keyTyped(char character) method in a set input processor gets called with when entering 🏗️ or similar codepoints.

As this issue relies on actual key input I don't how tests can be written for it.

#### Version of libGDX and/or relevant dependencies
ligGDX 1.12.0 with lwjgl3 backend.

#### Please select the affected platforms
- [X] Windows
Probably all platforms but I tested on Windows only.

**Comment by LobbyDivinus** [comment]
Apparently [Character.highSurrogate](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#highSurrogate-int-) and [Character.lowSurrogate](https://docs.oracle.com/javase/8/docs/api/java/lang/Character.html#lowSurrogate-int-) can be used to convert an integer codepoint to a char surrogate pair that would be compatible with String codepoint methods.

**Comment by tommyettinger** [comment]
Yeah... you're just scratching the surface of what makes emoji so incredibly challenging to get right. For example, this emoji "👨🏽‍❤️‍👨🏿" is made of the 12 `char`s `{'\uD83D', '\uDC68', '\uD83C', '\uDFFD', '\u200D', '\u2764', '\uFE0F', '\u200D', '\uD83D', '\uDC68', '\uD83C', '\uDFFF'}`. It actually involves multiple distinct emoji characters joined into one with a ZWJ (Zero-Width Joiner) character (which is of course zero-width and invisible). There are also skin tone modifiers applied to the two people. I'm not sure if Windows 11 can enter some of the most complex multi-part emoji into a form; Windows 10 can't to my knowledge. Windows 10 can specify a skin tone for one person, but not two people in one emoji char.

You might want to look at how [TextraTypist](https://github.com/tommyettinger/textratypist) handles this. It has a separate TextureAtlas that allows looking up the emoji bitmaps given their human-enterable names, which are Strings and not single chars (because that emoji I just showed is 12 chars, it really needs a String).

We can't change this easily without making keyTyped() return a String, which would be quite strange if each keystroke was its own String. That many String objects would also make Android's underpowered GC contemplate catching fire to end its suffering. There is probably some partial solution that gets words as they are entered, or something? I don't really know how native Android and iOS apps handle their autocorrect and emoji entry.

I think this should be closed as not planned, but I'd like to hear what other people think.

**Comment by LobbyDivinus** [comment]
Yes, emojis can be hard. My argument would be that given that glfw sends a 32 bit code point it may be beneficial to send that exact code point to the application using a surrogate pair to let the application handle it (assuming that the application builds up an input string by appending the characters it receives via keyTyped). From my testing the glfw callback gets called multiple times when multiple code points are in play for a single emoji. Issuing keyTyped multiple times then - dependent on the code point entered - should not be a big deal and wouldn't require to change existing applications and the interface would not change.

On the contrary, I would argue that the current approach is kind of broken as it removes information and by doing so effectively sends invalid characters to keyTyped. If issuing keyTyped potentially multple times to support surrogate pairs is not an option I'd propose to not send keyTyped in this case at all as the application cannot know if a character was part of a supplementary code point.

**Comment by LobbyDivinus** [comment]
In case someone else stumbles across this issue I was able to fix out without modifying libGDX by using inheritance and reflection.

Basically, I wrap the default char callback with a custom one that does handle 32bit codepoints in the following way:
```java
        charCallback = new GLFWCharCallback() {
            @Override
            public void invoke(long window, int codepoint) {
                if (Character.isSupplementaryCodePoint(codepoint)) {
                    defaultCharCallback.invoke(window, Character.highSurrogate(codepoint));
                    defaultCharCallback.invoke(window, Character.lowSurrogate(codepoint));
                } else {
                    defaultCharCallback.invoke(window, codepoint);
                }
            }
        };
```

The entry point is to override createInput() of Lwjgl3Application to let it return a custom DefaultLwgl3Input subclass that sets the own callback in windowHandleChanged() using `GLFW.glfwSetCharCallback(windowHandle, charCallback);`

Et voila, we can input emojis: 
<img width="88" alt="image" src="https://github.com/libgdx/libgdx/assets/7099530/93c82900-cf8d-4212-808e-ec38d15499ad">

**Comment by tommyettinger** [comment]
This is a clever solution! I don't see where it needs reflection, unless I'm missing something. That's not a bad thing! Graal Native Image has a hard time with reflection even in LWJGL3 projects. I think this can be done with just some additional anonymous class extensions for now... How does this handle emoji with skin color specifiers, such as 👨🏾‍💻 ? I'd guess it would be multiple sequential high-low pairs.

**Comment by LobbyDivinus** [comment]
Thanks! I needed reflection to get the old callback I called defaultCharCallback in my code as it is is not accessible in my subclass as it is package private.

Yes, the specifiers get reported to the application using multiple high low pairs. The game does not handle them correctly, yet. But in comparison to before it now receives all the data needed to handle them in the future 😁My main concern was that most emojis could not be entered at all and resulted in broken characters which is solved now.

**Comment by tommyettinger** [comment]
This might be appropriate to add to libGDX's input handling, since emoji are not going away anytime soon... I think we need to make sure something can render the emoji in libGDX, and I don't know if BitmapFont can.

**Comment by JamesXNelson** [comment]
So, just to add a little more flavor, `BitmapFont` does not support loading glyphs whose integer form > Character.MAX_VALUE (i.e. has a high/low surrogate).

```
int ch = Integer.parseInt(tokens.nextToken());
if (ch <= 0)
  missingGlyph = glyph;
else if (ch <= Character.MAX_VALUE)
  setGlyph(ch, glyph);
else
  continue;
```

In my case, I was able to generate fonts with large unicode values, like 🛠 (which I'm rendering with NotoSans2Symbol font).  The .fnt line looks like this:
```
char id=128736  x=49   y=0    width=19   height=17   xoffset=-1   yoffset=5    xadvance=18   page=0    chnl=0 
```

Seems like that check in BitmapFont could be doing `Character.isValidCodePoint(ch)`...  

I think I'll take a crack at making it work; the png and .fnt look good, just need to stop choking on valid chars.

**Comment by tommyettinger** [comment]
It's going to choke on ZWJ sequences even if it doesn't choke on individual 32-bit codepoints. Try entering "👨🏽‍❤️‍👨🏿", which only makes sense as a String because it contains 12 individual 16-bit `char` values... It cannot be represented as a `char`, nor an `int`. BitmapFont is not built to display a String as a single emoji glyph... If it could, we could also coincidentally handle complex scripts, like Khmer (which actually popped up in an issue once or twice) or any Devanagari-based Indic scripts, some of which have very complicated combining forms. There's some existing code on GitHub to support Arabic in libGDX, though it would need a lot of work to make it do things with emoji.

**Comment by JamesXNelson** [comment]
Is there some subset of characters that do fit into an upper surrogate and lower surrogate that could be supported, without having to handle all extreme cases?

**Comment by JamesXNelson** [comment]
...ya, my guy...  it is definitely possible to parse high/low surrogates and render at least some additional symbols.


It's... a touch shady, but I think altogether reliable in the end.  I wrote in a test w/ a sample image, and the test string "✓⌚✕🛠▶▼" totally works.

![Image](https://github.com/user-attachments/assets/9258206a-7853-47e9-aa70-a11deefea355)

It's only one actual "extra large" character, 🛠, but I consider it a win.  I can update the test w/ a different generated string of test characters from notosans2 if you like: https://fonts.google.com/noto/specimen/Noto+Sans+Symbols+2?query=symbol

Your test emoji character is not included in that font, but at least some of the less "extra" characters can actually be made to work.  I can easily generate most google fonts, and noto sans symbols was good enough for me to make "symbol buttons"...

I can sign the CLA and put together a PR if you like; I'd really rather not maintain my own copy.
Even if it can't handle every possible "character" a browser might, having a wider range of supported code points is nice at least :-)

**Comment by tommyettinger** [comment]
It might be useful for sure! What I wonder about are the emoji that aren't just monochrome, like what people usually expect when entering an emoji in a webpage (🤔). A BitmapFont can handle color just fine, that's one of the things Hiero is meant to do with things like effects, outlines, shadows and such. You're totally right about symbols for buttons; some parts of scene2d.ui are limited to text with no images, and this helps a lot in those places.

I wouldn't worry about the CLA, it hasn't been signed in a while, heh. https://github.com/libgdx/libgdx/blob/master/CONTRIBUTORS was last changed about 4 years ago!

**Comment by NathanSweet** [comment]
While it's obvious that PRs/contributions are subject to the libgdx license, sending the CLA is appreciated. CONTRIBUTORS is separate and should list all CLAs, though it may not.

**Comment by JamesXNelson** [comment]
Hi there.

Sorry I blanked on this for a while.

Do I just need to add my name to the CONTRIBUTORS file along w/ a PR?

**Comment by JamesXNelson** [comment]
fwiw, high/low surrogate support works great with creating font resources using emoji-containing fonts.  So, for things like buttons that you want driven by an icon-backed font, and don't mind it rendering monochrome, it works well enough.

AFAIK, support for a broad range of emoji would rely on huge font resources pre-rendering all supported images, unless there's some other way to render text w/ 'native' fonts I don't know about.

**Comment by tommyettinger** [comment]
Like https://github.com/tommyettinger/textratypist/blob/main/knownFonts/Twemoji.png ? It might be a good time to point out that the last time atlases were updated, the "fix" for Unicode image names didn't work, and I still need the workaround here: https://github.com/tommyettinger/textratypist/blob/main/src/main/java/com/github/tommyettinger/textra/KnownFonts.java#L4560-L4561 I vaguely remember the conversation, but I had provided the code I changed (one line only needed!) and that was ignored... https://github.com/libgdx/libgdx/blob/master/gdx/src/com/badlogic/gdx/graphics/g2d/TextureAtlas.java#L360 is still dependent on platform encoding to this day.
